docs/testing-principles.md

```diff
@@ -0,0 +1,200 @@
+d) survive any reorganization of the internals of the task
+
+The ability to run against real orgs is an important
+confidence-building measure when working with an environment
+that constantly changes like SAAS services.
+
+Using our testing infrastructure, you do so like this:
+
+```s
+$ pytest -k test_license_preflight --org qa
+```
+
+### How it Works
+
+What happens under the cover is that the `create_task` fixture knows whether
+the test suite is being run against a real org or not. If so, it passes
+an `sf_org` object connected to a real Salesforce org. If not, it passes
+a fake `sf_org` which will emulate a `qa_org`.
+
+In a perfect world, we would achieve a very high level of test coverage
+when every fake/mock is turned "off" and all code is running against
+real systems, just as we have high coverage when they are ruened on.
+
+The code paths which one might expect to be inaccessible in this mode are
+code for handling errors, service outages, latency and high scale.
+
+All other code should run against _real_ services or faked services
+interchangably, depending on the current step in the SDLC.
+
+### Component Tests
+
+One of the goals is to test our system as real users use it, which is why
+we use high-level APIs like `task()` instead of low level ones like
+`task._run_task()`
+
+That said, a system as complicated as Cumulus Framework cannot be tested only
+from the _end_ user's perspective. Cumulus Framework consists of many components.
+
+The "users" of these components are programmers on our team, and they should
+be well-tested, just as if they were documented external APIs.
+
+For example, the YAML merge algorithm is a component. The XML manipulator is
+a component. The package XML library SHOULD BE a component, etc. These should
+have test cases that ensure that their interfaces do not accidentally change.
+
+## The Hard Cases: The Awkward Squad
+
+In the lingo of functional programmers, our tests doesn't deal with the
+"awkward squad" of issues other tests need to deal with:
+
+-   a variety of org shapes
+-   error/exception conditions
+-   mutable state (in the org)
+-   delays, timestamps and other temporal aspects
+
+"Old fashioned" data faking may be needed to deal with these. The rest
+of this document outlines techninques for handling these hard cases.
+
+### Org shapes
+
+Cumulus Framework has testing infrastructure which can create custom
+orgs for tests:
+
+```
+@pytest.mark.vcr()
+@pytest.mark.org_shape("person_accounts", "config_person_accounts")
+def test_some_person_account_feature(self):
+    ...
+```
+
+This will run the `config_person_accounts` flow against the `person_accounts`
+org shape and run the test in the context of that org. At least that's
+the goal. This infrastructure has not been tested or used much yet (e.g. for
+actual person accounts testing).
+
+### Simulating Networking Errors with VCR
+
+It is possible to simulate networking errors by hand-editing VCR files.
```

Comment:
I am not persuaded that this strategy is superior to mocking - actually I think there's a case for the reverse. Applying this strategy results in VCR artifacts being a mix of automatically-generated and handwritten, muddying the story around how they're updated. And if the hand-edited cassettes need to be updated, the hand-editing process has to be repeated.

Is there a clear way to avoid that issue? Otherwise, I'm inclined to say that we still need mocks in some cases.